File: typescript/processor/src/processor.ts
/**
 * @docspec:module id="docspec-ts-processor" name="DocSpec TypeScript Processor"
 * @docspec:description "Main processor orchestrator for TypeScript codebases.
 *   Coordinates the full DocSpec v3 pipeline: TypeScript program creation,
 *   auto-discovery scan, package filtering, framework detection (NestJS, TypeORM,
 *   Prisma, Express), decorator/TSDoc/description reading, DSTI intent extraction
 *   with density scoring, domain extractor pipeline (security, config, observability,
 *   data stores, external deps, privacy, errors/events), coverage calculation,
 *   and final docspec.json assembly."
 * @docspec:boundary "TypeScript source code analysis"
 * @docspec:since "3.0.0"
 * @docspec:tags ["self-documented", "processor", "typescript", "pipeline"]
 */

import ts from "typescript";
import { AutoDiscoveryScanner } from "./scanner/auto-discovery.js";
import { PackageFilter } from "./scanner/package-filter.js";
import { DecoratorReader } from "./reader/decorator-reader.js";
import { TSDocReader } from "./reader/tsdoc-reader.js";
import { DescriptionInferrer } from "./reader/description-inferrer.js";
import { NestJSDetector } from "./framework/nestjs-detector.js";
import { TypeORMDetector } from "./framework/typeorm-detector.js";
import { PrismaDetector } from "./framework/prisma-detector.js";
import { IntentExtractor } from "./dsti/intent-extractor.js";
import { IntentDensityCalculator } from "./dsti/intent-density-calculator.js";
import { Serializer } from "./output/serializer.js";
import { CoverageCalculator } from "./metrics/coverage-calculator.js";
import { type DocSpecExtractor, type ExtractorContext, createExtractorModel } from "./extractor/extractor-interface.js";
import { SecurityExtractor } from "./extractor/security-extractor.js";
import { ConfigExtractor } from "./extractor/config-extractor.js";
import { ObservabilityExtractor } from "./extractor/observability-extractor.js";
import { DataStoreExtractor } from "./extractor/datastore-extractor.js";
import { ExternalDependencyExtractor } from "./extractor/external-dep-extractor.js";
import { PrivacyExtractor } from "./extractor/privacy-extractor.js";
import { ErrorEventExtractor } from "./extractor/error-event-extractor.js";
import { ExpressDetector } from "./extractor/express-detector.js";

export interface ProcessorConfig {
  tsConfigPath?: string;
  include?: string[];
  exclude?: string[];
  outputDir?: string;
  groupId?: string;
  artifactId?: string;
  version?: string;
}

export interface DocSpecOutput {
  docspec: "3.0.0";
  artifact: { groupId: string; artifactId: string; version: string; language: string; frameworks?: string[] };
  modules: any[];
  flows?: any[];
  contexts?: any[];
  dataModels?: any[];
  errors?: any[];
  events?: any[];
  security?: any;
  configuration?: any[];
  observability?: any;
  dataStores?: any[];
  externalDependencies?: any[];
  privacy?: any[];
  intentGraph?: { methods: any[] };
  discovery?: any;
}

export class DocSpecTSProcessor {
  private config: ProcessorConfig;
  private scanner: AutoDiscoveryScanner;
  private filter: PackageFilter;
  private decoratorReader: DecoratorReader;
  private tsdocReader: TSDocReader;
  private descriptionInferrer: DescriptionInferrer;
  private nestjsDetector: NestJSDetector;
  private typeormDetector: TypeORMDetector;
  private prismaDetector: PrismaDetector;
  private intentExtractor: IntentExtractor;
  private intentDensityCalculator: IntentDensityCalculator;
  private serializer: Serializer;
  private coverageCalculator: CoverageCalculator;
  private extractors: DocSpecExtractor[];

  constructor(config: ProcessorConfig) {
    this.config = config;
    this.scanner = new AutoDiscoveryScanner();
    this.filter = new PackageFilter(config.include ?? [], config.exclude ?? []);
    this.decoratorReader = new DecoratorReader();
    this.tsdocReader = new TSDocReader();
    this.descriptionInferrer = new DescriptionInferrer();
    this.nestjsDetector = new NestJSDetector();
    this.typeormDetector = new TypeORMDetector();
    this.prismaDetector = new PrismaDetector();
    this.intentExtractor = new IntentExtractor();
    this.intentDensityCalculator = new IntentDensityCalculator();
    this.serializer = new Serializer();
    this.coverageCalculator = new CoverageCalculator();
    this.extractors = [
      new SecurityExtractor(),
      new ConfigExtractor(),
      new ObservabilityExtractor(),
      new DataStoreExtractor(),
      new ExternalDependencyExtractor(),
      new PrivacyExtractor(),
      new ErrorEventExtractor(),
      new ExpressDetector(),
    ];
  }

  /** @docspec:intentional "Orchestrates the full DocSpec v3 pipeline: program creation, auto-discovery, filtering, framework detection, reading, DSTI, extraction, coverage, and JSON assembly" */
  process(): DocSpecOutput {
    // 1. Create TypeScript program
    const configPath = this.config.tsConfigPath ?? "tsconfig.json";
    const configFile = ts.readConfigFile(configPath, ts.sys.readFile);
    const parsedConfig = ts.parseJsonConfigFileContent(configFile.config, ts.sys, ".");
    const program = ts.createProgram(parsedConfig.fileNames, parsedConfig.options);
    const checker = program.getTypeChecker();

    // 2. Auto-discover exported types
    const discoveredTypes = this.scanner.scan(program, checker);

    // 3. Filter by package rules
    const filteredTypes = discoveredTypes.filter(t => this.filter.accept(t.fileName));

    // 4. Detect frameworks
    const frameworks: string[] = [];
    if (this.nestjsDetector.detect(program)) frameworks.push("nestjs");
    if (this.typeormDetector.detect(program)) frameworks.push("typeorm");
    if (this.prismaDetector.detect(program)) frameworks.push("prisma");

    // 5. Build modules from discovered types
    const modules: any[] = [];
    const moduleMap = new Map<string, any>();
    const intentMethods: any[] = [];

    for (const discovered of filteredTypes) {
      const sourceFile = discovered.sourceFile;
      const node = discovered.node;

      // Read decorators
      const decoratorMeta = this.decoratorReader.read(node);
      // Read TSDoc
      const tsdocMeta = this.tsdocReader.read(node, sourceFile);
      // Infer description if missing
      const description = tsdocMeta.description || this.descriptionInferrer.infer(discovered.name);

      // Determine module
      const moduleName = decoratorMeta.module?.name ?? this.inferModuleName(discovered.fileName);
      if (!moduleMap.has(moduleName)) {
        moduleMap.set(moduleName, { id: moduleName, name: moduleName, members: [] });
      }

      // Build member
      const member: any = {
        kind: discovered.kind,
        name: discovered.name,
        qualified: discovered.qualified,
        description,
        ...decoratorMeta.extra,
      };

      // Extract methods
      if (ts.isClassDeclaration(node)) {
        member.methods = this.extractMethods(node, sourceFile, checker, intentMethods, discovered.qualified);
        member.fields = this.extractFields(node, sourceFile, checker);
      }

      // Framework enrichment (only for class declarations)
      if (ts.isClassDeclaration(node)) {
        if (frameworks.includes("nestjs")) {
          this.nestjsDetector.enrich(node, member);
        }
        if (frameworks.includes("typeorm")) {
          this.typeormDetector.enrich(node, member);
        }
      }

      moduleMap.get(moduleName)!.members.push(member);
    }

    moduleMap.forEach(mod => modules.push(mod));

    // 6. Run v3 extractors (security, config, observability, data stores, etc.)
    const extractorModel = createExtractorModel();
    const userSourceFiles = program.getSourceFiles().filter(
      sf => !sf.isDeclarationFile && !sf.fileName.includes("node_modules"),
    );
    const extractorContext: ExtractorContext = {
      program,
      checker,
      sourceFiles: userSourceFiles,
      model: extractorModel,
    };

    for (const extractor of this.extractors) {
      if (extractor.isAvailable()) {
        extractor.extract(extractorContext);
      }
    }

    // 7. Compute ISD scores for intent methods
    for (const im of intentMethods) {
      if (im.intentSignals) {
        im.intentSignals.intentDensityScore = this.intentDensityCalculator.calculate(im.intentSignals);
      }
    }

    // 8. Calculate coverage
    this.coverageCalculator.reset();
    this.coverageCalculator.analyze(modules);
    const coverageReport = this.coverageCalculator.toReport(
      "auto",
      frameworks,
      this.config.include ?? [],
      this.config.exclude ?? [],
    );

    // 9. Build output
    const output: DocSpecOutput = {
      docspec: "3.0.0",
      artifact: {
        groupId: this.config.groupId ?? "unknown",
        artifactId: this.config.artifactId ?? "unknown",
        version: this.config.version ?? "0.0.0",
        language: "typescript",
        frameworks: frameworks.length > 0 ? frameworks : undefined,
      },
      modules,
      errors: extractorModel.errors.length > 0 ? extractorModel.errors : undefined,
      events: extractorModel.events.length > 0 ? extractorModel.events : undefined,
      security: extractorModel.security ?? undefined,
      configuration: extractorModel.configuration.length > 0 ? extractorModel.configuration : undefined,
      observability: extractorModel.observability ?? undefined,
      dataStores: extractorModel.dataStores.length > 0 ? extractorModel.dataStores : undefined,
      externalDependencies: extractorModel.externalDependencies.length > 0 ? extractorModel.externalDependencies : undefined,
      privacy: extractorModel.privacy.length > 0 ? extractorModel.privacy : undefined,
      intentGraph: intentMethods.length > 0 ? { methods: intentMethods } : undefined,
      discovery: coverageReport,
    };

    return output;
  }

  /** @docspec:deterministic */
  private inferModuleName(fileName: string): string {
    const parts = fileName.replace(/\\/g, "/").split("/");
    const srcIdx = parts.indexOf("src");
    if (srcIdx >= 0 && srcIdx + 1 < parts.length) {
      return parts[srcIdx + 1];
    }
    return "default";
  }

  /** @docspec:intentional "Extracts method metadata including TSDoc, parameters, return types, and DSTI intent signals from class declarations" */
  private extractMethods(node: ts.ClassDeclaration, sourceFile: ts.SourceFile, checker: ts.TypeChecker, intentMethods: any[], parentQualified: string): any[] {
    const methods: any[] = [];
    for (const member of node.members) {
      if (ts.isMethodDeclaration(member) && member.name) {
        const name = member.name.getText(sourceFile);
        const tsdoc = this.tsdocReader.readMember(member, sourceFile);
        const sig = checker.getSignatureFromDeclaration(member);
        const returnType = sig ? checker.typeToString(checker.getReturnTypeOfSignature(sig)) : "void";

        const method: any = {
          name,
          description: tsdoc.description || this.descriptionInferrer.infer(name),
          params: member.parameters.map(p => ({
            name: p.name.getText(sourceFile),
            type: p.type ? p.type.getText(sourceFile) : "any",
            required: !p.questionToken && !p.initializer,
          })),
          returns: { type: returnType },
        };

        // Extract intent signals
        const signals = this.intentExtractor.extractFromMethod(member, sourceFile);
        if (signals) {
          intentMethods.push({ qualified: `${parentQualified}#${name}`, intentSignals: signals });
        }

        methods.push(method);
      }
    }
    return methods;
  }

  /** @docspec:intentional "Extracts field names and types from class property declarations" */
  private extractFields(node: ts.ClassDeclaration, sourceFile: ts.SourceFile, checker: ts.TypeChecker): any[] {
    const fields: any[] = [];
    for (const member of node.members) {
      if (ts.isPropertyDeclaration(member) && member.name) {
        const name = member.name.getText(sourceFile);
        const type = member.type ? member.type.getText(sourceFile) : "any";
        fields.push({ name, type });
      }
    }
    return fields;
  }
}

File: rust/processor/src/processor.rs
//! Main processor orchestrator.
//!
//! @docspec:module id="docspec-rust-processor" name="DocSpec Rust Processor"
//! @docspec:description "Main processor orchestrator for Rust codebases.
//!   Coordinates the full DocSpec v3 pipeline: source file scanning,
//!   framework detection (Axum, Diesel, Serde), syn-based AST parsing to
//!   extract structs/enums/traits/functions/impl blocks, domain extractor
//!   pipeline (security, config, observability, data stores, external deps,
//!   privacy, errors/events), DSTI intent graph construction with density
//!   scoring, coverage calculation, and final docspec.json assembly."
//! @docspec:boundary "Rust source code analysis"
//! @docspec:since "3.0.0"
//! @docspec:tags ["self-documented", "processor", "rust", "pipeline"]
//!
//! Coordinates the full DocSpec pipeline:
//! 1. Scan source files
//! 2. Detect frameworks
//! 3. Parse AST and extract modules/members
//! 4. Run domain extractors (security, config, observability, etc.)
//! 5. Build DSTI intent graph with density scores
//! 6. Calculate documentation coverage metrics
//! 7. Assemble final docspec.json

use std::path::PathBuf;
use serde_json::Value;
use crate::scanner;
use crate::reader::{doc_comment_reader, description_inferrer};
use crate::framework::{axum_detector, diesel_detector, serde_detector};
use crate::dsti::{intent_extractor, intent_density_calculator, naming_analyzer};
use crate::extractor;
use crate::metrics::coverage;

/// @docspec:boundary "Orchestrates the complete DocSpec v3 pipeline from Rust source to docspec.json"
/// @docspec:intentional "Coordinates framework detection, AST parsing, domain extraction, DSTI analysis, and coverage calculation into a single pipeline"
pub struct RustProcessor {
    pub source_dir: PathBuf,
    pub output_dir: PathBuf,
}

impl RustProcessor {
    /// @docspec:method { since: "3.0.0" }
    /// @docspec:deterministic
    pub fn new(source_dir: PathBuf, output_dir: PathBuf) -> Self {
        Self { source_dir, output_dir }
    }

    /// @docspec:method { since: "3.0.0" }
    /// @docspec:intentional "Executes the full 7-step DocSpec pipeline: framework detection, domain extraction, AST parsing, intent graph construction, spec assembly, extractor merge, and coverage calculation"
    /// @docspec:preserves "source files are read-only; no mutations to the input codebase"
    pub fn process(&self) -> Result<Value, String> {
        let files = scanner::scan(&self.source_dir)?;
        let mut modules = Vec::new();
        let mut intent_methods = Vec::new();
        let mut frameworks: Vec<String> = Vec::new();
        let mut inferred_count: u32 = 0;

        // --- Step 1: Detect frameworks ---
        if axum_detector::detect(&self.source_dir) { frameworks.push("axum".to_string()); }
        if diesel_detector::detect(&self.source_dir) { frameworks.push("diesel".to_string()); }
        if serde_detector::detect(&self.source_dir) { frameworks.push("serde".to_string()); }

        // --- Step 2: Run domain extractors ---
        let extractor_ctx = extractor::run_all(&files, &self.source_dir);

        // Merge extractor-detected frameworks
        for fw in &extractor_ctx.frameworks {
            if !frameworks.contains(fw) {
                frameworks.push(fw.clone());
            }
        }

        // --- Step 3: Parse AST and build modules ---
        for file_info in &files {
            let syntax = syn::parse_file(&file_info.source)
                .map_err(|e| format!("Parse error in {}: {}", file_info.path.display(), e))?;

            let mut members = Vec::new();
            for item in &syntax.items {
                match item {
                    syn::Item::Struct(s) => {
                        let name = s.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let (description, was_inferred) = get_description(&s.attrs, &name);
                        if was_inferred { inferred_count += 1; }

                        let fields: Vec<Value> = s.fields.iter().filter_map(|f| {
                            f.ident.as_ref().map(|ident| {
                                let ty = &f.ty;
                                let mut field_json = serde_json::json!({
                                    "name": ident.to_string(),
                                    "type": quote::quote!(#ty).to_string(),
                                });
                                if let Some(doc) = doc_comment_reader::read_doc_comment(&f.attrs) {
                                    field_json.as_object_mut().unwrap()
                                        .insert("description".to_string(), Value::String(doc));
                                }
                                field_json
                            })
                        }).collect();

                        members.push(serde_json::json!({
                            "kind": "struct",
                            "name": name,
                            "qualified": qualified,
                            "description": description,
                            "fields": fields,
                        }));
                    }
                    syn::Item::Enum(e) => {
                        let name = e.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let (description, was_inferred) = get_description(&e.attrs, &name);
                        if was_inferred { inferred_count += 1; }

                        let values: Vec<String> = e.variants.iter()
                            .map(|v| v.ident.to_string()).collect();

                        members.push(serde_json::json!({
                            "kind": "enum",
                            "name": name,
                            "qualified": qualified,
                            "description": description,
                            "values": values,
                        }));
                    }
                    syn::Item::Trait(t) => {
                        let name = t.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let (description, was_inferred) = get_description(&t.attrs, &name);
                        if was_inferred { inferred_count += 1; }

                        members.push(serde_json::json!({
                            "kind": "trait",
                            "name": name,
                            "qualified": qualified,
                            "description": description,
                        }));
                    }
                    syn::Item::Fn(f) => {
                        let name = f.sig.ident.to_string();
                        let qualified = format!("{}::{}", file_info.module, name);
                        let (description, was_inferred) = get_description(&f.attrs, &name);
                        if was_inferred { inferred_count += 1; }

                        // Extract intent signals (original DSTI)
                        if let Some(signals) = intent_extractor::extract_from_fn(&f.sig, &f.block) {
                            intent_methods.push(serde_json::json!({
                                "qualified": qualified,
                                "intentSignals": signals,
                            }));
                        }

                        // Run naming analyzer for richer semantics (used in DSTI)
                        let _name_semantics = naming_analyzer::analyze(&name);

                        members.push(serde_json::json!({
                            "kind": "function",
                            "name": name,
                            "qualified": qualified,
                            "description": description,
                        }));
                    }
                    syn::Item::Impl(imp) => {
                        if let syn::Type::Path(tp) = imp.self_ty.as_ref() {
                            let type_name = tp.path.segments.last()
                                .map(|s| s.ident.to_string())
                                .unwrap_or_default();

                            for impl_item in &imp.items {
                                if let syn::ImplItem::Fn(method) = impl_item {
                                    let method_name = method.sig.ident.to_string();
                                    let method_qualified = format!("{}::{}#{}", file_info.module, type_name, method_name);

                                    if let Some(signals) = intent_extractor::extract_from_fn(&method.sig, &method.block) {
                                        intent_methods.push(serde_json::json!({
                                            "qualified": method_qualified,
                                            "intentSignals": signals,
                                        }));
                                    }
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }

            if !members.is_empty() {
                modules.push(serde_json::json!({
                    "id": file_info.module,
                    "name": file_info.module,
                    "members": members,
                }));
            }
        }

        // --- Step 4: Build intent graph with density scores ---
        let intent_graph = if intent_methods.is_empty() {
            None
        } else {
            let mut ig = serde_json::json!({"methods": intent_methods});
            intent_density_calculator::enrich_intent_graph(&mut ig);
            Some(ig)
        };

        // --- Step 5: Assemble spec ---
        let mut spec = serde_json::json!({
            "docspec": "3.0.0",
            "artifact": {
                "groupId": "unknown",
                "artifactId": "unknown",
                "version": "0.0.0",
                "language": "rust",
            },
            "modules": modules,
        });

        // Scope mutable borrow so we can immutably borrow `spec` for coverage later
        {
            let spec_map = spec.as_object_mut().unwrap();

            // Add frameworks
            if !frameworks.is_empty() {
                spec_map["artifact"].as_object_mut().unwrap()
                    .insert("frameworks".to_string(), serde_json::json!(frameworks));
            }

            // Add intent graph
            if let Some(ig) = intent_graph {
                spec_map.insert("intentGraph".to_string(), ig);
            }

            // --- Step 6: Merge extractor results ---
            if let Some(security) = &extractor_ctx.security {
                spec_map.insert("security".to_string(), security.clone());
            }
            if !extractor_ctx.configuration.is_empty() {
                spec_map.insert("configuration".to_string(), Value::Array(extractor_ctx.configuration.clone()));
            }
            if let Some(observability) = &extractor_ctx.observability {
                spec_map.insert("observability".to_string(), observability.clone());
            }
            if !extractor_ctx.data_stores.is_empty() {
                spec_map.insert("dataStores".to_string(), Value::Array(extractor_ctx.data_stores.clone()));
            }
            if !extractor_ctx.external_dependencies.is_empty() {
                spec_map.insert("externalDependencies".to_string(), Value::Array(extractor_ctx.external_dependencies.clone()));
            }
            if !extractor_ctx.privacy.is_empty() {
                spec_map.insert("privacy".to_string(), Value::Array(extractor_ctx.privacy.clone()));
            }
            if !extractor_ctx.errors.is_empty() {
                spec_map.insert("errors".to_string(), Value::Array(extractor_ctx.errors.clone()));
            }
            if !extractor_ctx.events.is_empty() {
                spec_map.insert("events".to_string(), Value::Array(extractor_ctx.events.clone()));
            }
        } // spec_map dropped here

        // --- Step 7: Calculate coverage ---
        let mut stats = coverage::analyze(&spec);
        stats.inferred_descriptions = inferred_count;

        let scanned = vec![self.source_dir.to_string_lossy().to_string()];
        let discovery = coverage::to_discovery_model(
            &stats,
            "auto",
            &frameworks,
            &scanned,
            &[],
        );
        spec.as_object_mut().unwrap()
            .insert("discovery".to_string(), discovery);

        Ok(spec)
    }
}

/// Get description from doc comments or infer it from the name.
/// Returns (description, was_inferred).
///
/// @docspec:deterministic
/// @docspec:intentional "Falls back to name-based inference when no doc comment exists, ensuring every item has a description"
fn get_description(attrs: &[syn::Attribute], name: &str) -> (String, bool) {
    match doc_comment_reader::read_doc_comment(attrs) {
        Some(doc) => (doc, false),
        None => (description_inferrer::infer(name), true),
    }
}
